src/base/locked_queue_index.h

```diff
@@ -19,46 +29,112 @@ class LockedQueue<T>::Index {
   ui64 Size() const THREAD_UNSAFE {
     ui64 index_size = 0u;
     for (const auto& shard : index_) {
-      index_size += shard.size();
+      index_size += shard.tasks.size();
     }
     return index_size;
   }
 
-  void Put(QueueIterator it, ui32 shard) {
-    // TODO(ilezhankin): describe rationale for this place.
-    if (shard >= index_.size()) {
-      index_.resize(shard + 1);
+  void Put(QueueIterator it, const ui32 shard) {
+    EnsureShardExists(shard);
+    reverse_index_[&*it] = std::make_pair(
+        index_[shard].tasks.insert(index_[shard].tasks.end(), it), shard);
+  }
+
+  bool ShardIsEmpty(const ui32 shard) THREAD_UNSAFE {
+    DCHECK(shard < index_.size());
+    return index_[shard].tasks.empty();
+  }
+
+  void WaitShardFor(const ui32 shard, UniqueLock& lock,
+                    const Seconds& timeout) THREAD_UNSAFE {
+    EnsureShardExists(shard);
+    DCHECK(shard < index_.size());
+    index_[shard].pop_condition_.wait_for(lock, timeout);
+  }
+
+  template <typename Pred>
+  void WaitDefaultShard(UniqueLock& lock, const Pred& pred) THREAD_UNSAFE {
+    index_[LockedQueue<T>::DEFAULT_SHARD].pop_condition_.wait(lock, pred);
+  }
+
+  void NotifyShard(const ui32 shard) THREAD_SAFE {
+    DCHECK(shard < index_.size());
+    index_[shard].pop_condition_.notify_one();
```

Comment:
Since this method pretends to be thread-safe, how can we guarantee that the `size()` is changed atomically with `index_[shard]`?